src/lua_console.cpp

```diff
@@ -0,0 +1,82 @@
+#include "lua_console.h"
+
+#include "catalua.h"
+#include "catacharset.h"
+
+lua_console::lua_console() : cWin( newwin( lines, width, 0, 0 ) ), iWin( newwin( 1, width,lines, 0 ) )
+{
+}
+
+void lua_console::dispose()
+{
+    werase( cWin );
+    werase( iWin );
+    delwin( cWin );
+    delwin( iWin );
+}
+
+std::string lua_console::get_input()
+{
+    long key = 0;
+    int pos = -1;
+    return string_input_win( iWin, "", width, 0, 0, width, true, key, pos, "LUA", 0, lines, true, false, callbacks );
+}
+
+void lua_console::draw()
+{
+    werase( cWin );
+
+    // Some juggling to make sure text is aligned with the bottom of the console.
+    int stack_size = text_stack.size() - scroll;
```

Comment:
At the least bundle up your algorithm and put it with the other foldstring
functions so they can be reused and maintained together, ideally update an
existing method to provide the functionality you need.